File: terraform/src/main/kotlin/com/hyfly/tf/actuator/entity/constants/TfCommand.java
package com.hyfly.tf.actuator.entity.constants;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.experimental.Accessors;
import org.apache.commons.lang3.StringUtils;

import java.util.LinkedList;
import java.util.List;

@Data
@Accessors(chain = true)
@NoArgsConstructor
@AllArgsConstructor
public class TfCommand {

    public static final String _VAR = "-var=";

    public static final String _PLUGIN_DIR = "-plugin-dir=";

    /**
     * 用于 ProcessActuator 是否按行解析
     */
    private Boolean isLineParse;

    /**
     * 基础命令, 例如: terraform init
     */
    private LinkedList<String> baseCommand;

    /**
     * 插件目录, 一般只给 terraform init 使用,例如: terraform init -plugin-dir=/path/to/plugin
     */
    private String pluginDir;

    /**
     * 变量, 例如: terraform plan -var="key=value"
     */
    private List<String> variables;

    public TfCommand(Boolean isLineParse, LinkedList<String> baseCommand) {
        this.isLineParse = isLineParse;
        this.baseCommand = baseCommand;
    }

    public TfCommand appendVariable(String var) {
        if (this.variables == null) {
            this.variables = new LinkedList<>();
        }
        this.variables.add(var);
        return this;
    }

    public TfCommand appendVariables(LinkedList<String> vars) {
        if (this.variables == null) {
            this.variables = new LinkedList<>();
        }
        this.variables.addAll(vars);
        return this;
    }

    public String[] getCommand() {
        LinkedList<String> command = new LinkedList<>(this.baseCommand);

        if (StringUtils.isNotBlank(this.pluginDir)) {
            command.add(_PLUGIN_DIR + this.pluginDir);
        }

        List<String> vars = this.variables;
        if (vars != null && !vars.isEmpty()) {
            for (String var : vars) {
                command.add(_VAR + var);
            }
        }

        return command.toArray(new String[0]);
    }
}

File: terraform/src/main/kotlin/com/hyfly/tf/processor/entity/bo/TfCommand.kt
package com.hyfly.tf.processor.entity.bo

import org.apache.commons.lang3.StringUtils
import java.util.*

data class TfCommand(

    /**
     * 用于 ProcessActuator 是否按行解析
     */
    var isLineParse: Boolean = false,

    /**
     * 基础命令, 例如: terraform init
     */
    var baseCommand: LinkedList<String> = LinkedList(),

    /**
     * 插件目录, 一般只给 terraform init 使用,例如: terraform init -plugin-dir=/path/to/plugin
     */
    var pluginDir: String? = null,

    /**
     * 变量, 例如: terraform plan -var="key=value"
     */
    var variables: MutableList<String>? = null
) {
    companion object {
        const val _VAR = "-var="
        const val _PLUGIN_DIR = "-plugin-dir="
    }

    constructor(isLineParse: Boolean, baseCommand: LinkedList<String>) : this(isLineParse, baseCommand, null, null)

    fun appendVariable(var_: String): TfCommand {
        if (variables == null) {
            variables = LinkedList()
        }
        variables!!.add(var_)
        return this
    }

    fun appendVariables(vars: LinkedList<String>): TfCommand {
        if (variables == null) {
            variables = LinkedList()
        }
        variables!!.addAll(vars)
        return this
    }

    fun getCommand(): Array<String> {
        val command = LinkedList(baseCommand)

        if (StringUtils.isNotBlank(pluginDir)) {
            command.add(_PLUGIN_DIR + pluginDir)
        }

        variables?.let { vars ->
            if (vars.isNotEmpty()) {
                vars.forEach { s ->
                    command.add(_VAR + s)
                }
            }
        }

        return command.toTypedArray()
    }

}
